Feature/upf multicast router

joseprrm opened the pull request:
Use Linux Multicast Forwarding Cache to automatically create multicast routes towards the ogstun in the MB-UPF. With this, we no longer need to manually create multicast routes using smcroute.

The multicast router is configured in the upf configuration under the mbs section:
```
  mbs:
    multicast_router:
      input_interface: eth0
      output_interface: ogstun
      activate: true
```

- the multicast router is only used when "activate: true" is configured.
- input_interface is the interface that receives the multicast traffic from the AF. 
    - if it is not configured and the host only has one interface (other than "lo" or "ogstun"), it will automatically use that interface.
- output_interface is the tun interface used to send the traffic towards the UEs. 
    - if it is not configured, it will try to use "ogstun" by default.

Because of the way Linux Multicast Forwarding Cache works, only one MB-UPF could configure multicast routes in one network namespace.

# Example
The MB-UPF is configured as follows:
```
  mbs:
    multicast_router:
      input_interface: eth0
      output_interface: ogstun
      activate: true
```

Send the following request from the AF to the MB-SMF:
```
curl --silent --http2-prior-knowledge --request POST --header 'Content-Type: application/json' \
--data '{ "mbsSession": { "ssm": { "sourceIpAddr": { "ipv4Addr": "10.33.33.3" }, "destIpAddr": { "ipv4Addr": "239.0.0.25" } },"tmgiAllocReq": true, "serviceType":"BROADCAST" }}' \
mb-smf.open5gs.org:80/nmbsmf-mbssession/v1/mbs-sessions
```
- "sourceIpAddr" is the origin IP that is going to send the multicast traffic, usually the AF.
- "destIpAddr" is the multicast group.

After the request is received, the MB-UPF will automatically configure a route that will forward traffic to the 239.0.0.25 multicast group coming from 10.33.33.3 on the eth0 interface, towards ogstun interface.

joseprrm committed:
Add multicast router in UPF

Add support for UPF to join multicast groups and setup multicast routes
using Linux Multicast Forwarding Cache.

joseprrm committed:
Add multicast route for every PDR

davidjwbbc commented on src/upf/n4mb-handler.c:
```diff
@@ -300,6 +302,29 @@ void upf_n4mb_handle_session_establishment_request(
     }
 
+
+    if(upf_self()->mbs_multicastrouter_activate){
```
The multicast routes do not need setting up if a UDP tunnel endpoint has been requested (no routing required). This code should not be executed if `req->create_traffic_endpoint.presence` is `true`.

```suggestion
    if (upf_self()->mbs_multicastrouter_activate && !req->create_traffic_endpoint.presence) {
```

joseprrm replied:
Solved in 255f4f243acb777125eaf562bc9aec3feb28089b

jordijoangimenez replied:
Hello, the issue of my code not executing is here as req->create_traffic_endpoint.presence is true, therefore false and nothing below the if is executed.

How is req->create_traffic_endpoint.presence configured to false? (which parameter in the yaml or in the call?). Just for the sake of testing all.

davidjwbbc replied:
`req->create_traffic_endpoint.presence` is true if the create_traffic_endpoint PFCP IE is present in the message from the MB-SMF. The MB-SMF only adds this IE when either the `ingressTunAddrReq` flag is set or if the `ingressTunAddr` address is set in the MBS Session. Otherwise the IE is not added and so the presence flag is not set.

jordijoangimenez replied:
Thanks @davidjwbbc , then the actual flag is "ingressTunAddrReq": true from the create MBS JSON and not the multicast_router: activate: true in the yml

davidjwbbc commented on src/upf/n4mb-handler.c:
```diff
@@ -300,6 +302,29 @@ void upf_n4mb_handle_session_establishment_request(
+            inet_ntop(AF_INET, &ip_multicast, ip_multicast_string, sizeof ip_multicast_string);
+            inet_ntop(AF_INET, &ip_source, ip_source_string, sizeof ip_source_string);
+
+            if(OGS_ERROR == multicastrouter_add_route_and_join(ip_source_string, ip_multicast_string)){
```
This seems a waste of effort by the CPU to convert IP addresses into strings and then have the `multicastrouter_add_route_and_join` function immediately convert them back into IP addresses. You should probably just pass `in_addr_t` types for the addresses as noted in my previous comment on the *multicastrouter.c* file.

joseprrm replied:
I am going to leave it as it is, because to print the routes in the logs I need the string anyway.

davidjwbbc commented on lib/multicastrouter/multicastrouter.c:
```diff
@@ -0,0 +1,183 @@
+    }
+}
+
+int multicastrouter_add_route_and_join(const char *ip_source_string, const char *ip_multicast_string){
+    int error;
+
+    // ADD MULTICAST ROUTE
+    // we need source ip address, multicast ip address, input interface(IIF), output interface (OIF)
+    in_addr_t source = inet_addr(ip_source_string);
+    in_addr_t group = inet_addr(ip_multicast_string);
```
Why does this function take strings when better primitive types, such as `in_addr_t`, exist? It seems a waste to make calling functions convert the IP addresses into strings to call this function only to have the values converted back into `in_addr_t` types.

davidjwbbc commented on src/upf/n4mb-handler.c:
```diff
@@ -300,6 +302,29 @@ void upf_n4mb_handle_session_establishment_request(
+            ogs_assert(pdr);
+            // get the ips from the PDR
+            uint32_t ip_multicast = pdr->ip_multicast_addressing_info.ip_multicast_address.s_ipv4_addr;
+            uint32_t ip_source    = pdr->ip_multicast_addressing_info.source_ip_address.ipv4_addr;
```
This code only deals with IPv4 multicast addresses, what happens if an attempt to configure IPv6 multicast is made? I see no protection to not execute this code when only IPv6 is used. You should be checking the `pdr->ip_multicast_addressing_info.ip_multicast_address.v4` and `pdr->ip_multicast_addressing_info.source_ip_address.v4` flags before proceeding, to ensure IPv4 addresses are present for this PDR.

davidjwbbc replied:
There's a separate question about whether to implement IPv6 address support too.

joseprrm replied:
Thanks for pointing this out, I added the checks in 288efc192d31045184407496333eb5b5dedb8512
I also added a warning saying that the multicast router doesn't implement ipv6. I am going to leave it unimplemented for the moment, since I am only working with ipv4.

jordijoangimenez commented:
Some testing reproducing the tutorial of the MBSTF with this new MB-UPF and beyond the remarks from @davidjwbbc on the code.

I've configured the UPF yaml according to the instructions from @joseprrm in the PR.

In my case:

```
  mbs:
    udptunnel:
       address: 127.0.0.7
       port: 0
    multicast_router:
       input_interface: eno1
       output_interface: ogstun
       activate: true
```
I've used this MB-UPF to play with the MBSTF tutorial and it works (I see the FLUTE packets coming from 127.0.0.1 to 232.0.0.1

```
fivegmag@fivegmag:~/open5gs_mbs_josep$ sudo -E install/bin/open5gs-upfd 
Open5GS daemon v2.6.4-563-g943b0a6

01/20 16:49:25.113: [app] INFO: Configuration: '/home/fivegmag/open5gs_mbs_josep/install/etc/open5gs/upf.yaml' (../lib/app/ogs-init.c:144)
01/20 16:49:25.113: [app] INFO: File Logging: '/home/fivegmag/open5gs_mbs_josep/install/var/log/open5gs/upf.log' (../lib/app/ogs-init.c:147)
01/20 16:49:25.176: [metrics] INFO: metrics_server() [http://127.0.0.7]:9090 (../lib/metrics/prometheus/context.c:299)
01/20 16:49:25.176: [pfcp] INFO: pfcp_server() [127.0.0.7]:8805 (../lib/pfcp/path.c:30)
01/20 16:49:25.176: [gtp] INFO: gtp_server() [127.0.0.7]:2152 (../lib/gtp/path.c:30)
01/20 16:49:25.177: [upf] INFO: Initializing multicast router using eno1 as input interface and ogstun as output interface (../src/upf/multicastrouter-path.c:115)
01/20 16:49:25.177: [app] INFO: UPF initialize...done (../src/upf/app.c:31)
01/20 16:49:32.012: [pfcp] INFO: ogs_pfcp_connect() [127.0.0.4]:8805 (../lib/pfcp/path.c:61)
01/20 16:49:32.013: [upf] INFO: PFCP associated [127.0.0.4]:8805 (../src/upf/pfcp-sm.c:186)
01/20 16:49:42.276: [upf] INFO: [Added] Number of MBS Sessions in UPF is now 1 (../src/upf/context.c:1097)
01/20 16:49:42.276: [gtp] INFO: gtp_connect() [239.0.0.4]:2152 (../lib/gtp/path.c:60)
01/20 16:49:42.276: [upf] INFO: Multicast route created: 127.0.0.1 to 232.0.0.1 (../src/upf/n4mb-handler.c:322)
01/20 16:49:42.276: [upf] INFO: Join multicast group: 232.0.0.1 (../src/upf/n4mb-handler.c:323)
```

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/9908a913-1753-4e28-8f57-fffab2d1e835" />

jordijoangimenez commented:
I would recommend editing the upf.yaml to add an example for this configuration:

```
  mbs:
    udptunnel:
       address: 127.0.0.7
       port: 0
    multicast_router:
       input_interface: eno1
       output_interface: ogstun
       activate: true
``

davidjwbbc commented:
@jordijoangimenez wrote:
> I would recommend editing the upf.yaml to add an example for this configuration:
> 
> ```
>   mbs:
>     udptunnel:
>        address: 127.0.0.7
>        port: 0
>     multicast_router:
>        input_interface: eno1
>        output_interface: ogstun
>        activate: true
> ```

Would this be better as a comment in the YAML configuration so that it is not set as default? If the system is already running `mrouted` or the user has already used `smcroute`, then this is not necessary. Also the input interface may have a different name in another user's setup.

jordijoangimenez commented:
Agree, suggest adding these lines as #

dsilhavy commented:
@joseprrm Can you please check the comments by @davidjwbbc and @jordijoangimenez

joseprrm committed:
multicastrouter: only add routes for IPv4

joseprrm committed:
Do not use multicast router if udp tunneling is used

joseprrm committed:
Add example of multicast router configuration in UPF configuration

joseprrm commented:
@davidjwbbc @jordijoangimenez 

cc15b45556d69f6598b8bbff804a8385c6ed625c I added the commented configuration in `./configs/open5gs/upf.yaml.in`

I addressed the other comments by David in their own thread.

jordijoangimenez commented:
Thank you, @joseprrm ! We will check again.

jordijoangimenez commented on src/upf/n4mb-handler.c:
```diff
@@ -307,20 +307,26 @@ void upf_n4mb_handle_session_establishment_request(
-                ogs_info("Multicast route created: %s to %s", ip_source_string, ip_multicast_string);
-                ogs_info("Join multicast group: %s", ip_multicast_string);
+
+            if (pdr->ip_multicast_addressing_info.source_ip_address.v4 || pdr->ip_multicast_addressing_info.ip_multicast_address.v4) {
```
@joseprrm we have an issue somewhere here I think.

In your previous code I could see this
```
01/30 19:18:13.694: [gtp] INFO: gtp_server() [127.0.0.7]:2152 (../lib/gtp/path.c:30)
01/30 19:18:13.695: [upf] INFO: Initializing multicast router using eno1 as input interface and ogstun as output interface (../src/upf/multicastrouter-path.c:115)
01/30 19:18:13.695: [app] INFO: UPF initialize...done (../src/upf/app.c:31)
01/30 19:18:14.496: [pfcp] INFO: ogs_pfcp_connect() [127.0.0.4]:8805 (../lib/pfcp/path.c:61)
01/30 19:18:14.496: [upf] INFO: PFCP associated [127.0.0.4]:8805 (../src/upf/pfcp-sm.c:186)
01/30 19:18:31.138: [upf] INFO: [Added] Number of MBS Sessions in UPF is now 1 (../src/upf/context.c:1097)
01/30 19:18:31.138: [gtp] INFO: gtp_connect() [239.0.0.4]:2152 (../lib/gtp/path.c:60)
01/30 19:18:31.139: [upf] INFO: Multicast route created: 127.0.0.1 to 232.0.0.1 (../src/upf/n4mb-handler.c:322)
01/30 19:18:31.139: [upf] INFO: Join multicast group: 232.0.0.1 (../src/upf/n4mb-handler.c:323)
```

In your new code these lines are never shown (I suspect not executed):
```
ogs_info("Multicast route created: %s to %s", ip_source_string, ip_multicast_string);
                ogs_info("Join multicast group: %s", ip_multicast_string);
```

jordijoangimenez commented:
I'll try check more on Monday. Another question I have is if leaving this configs like below makes the router to be disregarded no matter if active is set to true because the udptunner prevails.

mbs: udptunnel: address: 127.0.0.7 port: 0 multicast_router: input_interface: eno1 output_interface: ogstun activate: true

Just pointing this out in case the if is not executed just because it should not be.

I see the log about the router being created though.

jordijoangimenez commented:
I had to test this before sleeping. Done on a clean VM.

This is the result with the checkout cc15b45
```
jordi@jordi-VMware-Virtual-Platform:~/open5gs$ sudo -E install/bin/open5gs-upfd
Open5GS daemon v2.6.4-566-gcc15b45

01/31 00:26:12.802: [app] INFO: Configuration: '/home/jordi/open5gs/install/etc/open5gs/upf.yaml' (../lib/app/ogs-init.c:144)
01/31 00:26:12.802: [app] INFO: File Logging: '/home/jordi/open5gs/install/var/log/open5gs/upf.log' (../lib/app/ogs-init.c:147)
01/31 00:26:12.851: [metrics] INFO: metrics_server() [http://127.0.0.7]:9090 (../lib/metrics/prometheus/context.c:299)
01/31 00:26:12.851: [pfcp] INFO: pfcp_server() [127.0.0.7]:8805 (../lib/pfcp/path.c:30)
01/31 00:26:12.851: [gtp] INFO: gtp_server() [127.0.0.7]:2152 (../lib/gtp/path.c:30)
01/31 00:26:12.853: [upf] INFO: Initializing multicast router using ens33 as input interface and ogstun as output interface (../src/upf/multicastrouter-path.c:115)
01/31 00:26:12.853: [app] INFO: UPF initialize...done (../src/upf/app.c:31)
01/31 00:26:15.405: [pfcp] INFO: ogs_pfcp_connect() [127.0.0.4]:8805 (../lib/pfcp/path.c:61)
01/31 00:26:15.405: [upf] INFO: PFCP associated [127.0.0.4]:8805 (../src/upf/pfcp-sm.c:186)
01/31 00:26:25.041: [upf] INFO: [Added] Number of MBS Sessions in UPF is now 1 (../src/upf/context.c:1097)
01/31 00:26:25.042: [gtp] INFO: gtp_connect() [239.0.0.4]:2152 (../lib/gtp/path.c:60)
```

This is the result with the checkout 943b0a6

```
jordi@jordi-VMware-Virtual-Platform:~/open5gs$ sudo -E install/bin/open5gs-upfd
Open5GS daemon v2.6.4-563-g943b0a6

01/31 00:35:20.062: [app] INFO: Configuration: '/home/jordi/open5gs/install/etc/open5gs/upf.yaml' (../lib/app/ogs-init.c:144)
01/31 00:35:20.062: [app] INFO: File Logging: '/home/jordi/open5gs/install/var/log/open5gs/upf.log' (../lib/app/ogs-init.c:147)
01/31 00:35:20.108: [metrics] INFO: metrics_server() [http://127.0.0.7]:9090 (../lib/metrics/prometheus/context.c:299)
01/31 00:35:20.108: [pfcp] INFO: pfcp_server() [127.0.0.7]:8805 (../lib/pfcp/path.c:30)
01/31 00:35:20.108: [gtp] INFO: gtp_server() [127.0.0.7]:2152 (../lib/gtp/path.c:30)
01/31 00:35:20.109: [upf] INFO: Initializing multicast router using ens33 as input interface and ogstun as output interface (../src/upf/multicastrouter-path.c:115)
01/31 00:35:20.111: [app] INFO: UPF initialize...done (../src/upf/app.c:31)
01/31 00:35:27.413: [pfcp] INFO: ogs_pfcp_connect() [127.0.0.4]:8805 (../lib/pfcp/path.c:61)
01/31 00:35:27.413: [upf] INFO: PFCP associated [127.0.0.4]:8805 (../src/upf/pfcp-sm.c:186)
01/31 00:35:30.307: [upf] INFO: [Added] Number of MBS Sessions in UPF is now 1 (../src/upf/context.c:1097)
01/31 00:35:30.307: [gtp] INFO: gtp_connect() [239.0.0.4]:2152 (../lib/gtp/path.c:60)
01/31 00:35:30.308: [upf] INFO: Multicast route created: 127.0.0.1 to 232.0.0.1 (../src/upf/n4mb-handler.c:322)
01/31 00:35:30.309: [upf] INFO: Join multicast group: 232.0.0.1 (../src/upf/n4mb-handler.c:323)
```

jordijoangimenez commented:
Bringing the remaining issue up here: https://github.com/5G-MAG/open5gs/pull/34#discussion_r2749661078

jordijoangimenez reviewed (approved):
Ready to merge. Tested with the request on false.
